core/src/avm1/globals/rectangle.rs

```diff
@@ -0,0 +1,1156 @@
+            context,
+        )?;
+        this.set(
+            "height",
+            args.get(3).unwrap_or(&Value::Undefined).to_owned(),
+            avm,
+            context,
+        )?;
+    }
+
+    Ok(Value::Undefined.into())
+}
+
+fn to_string<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    let x = this
+        .get("x", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+    let y = this
+        .get("y", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+    let width = this
+        .get("width", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+    let height = this
+        .get("height", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+
+    Ok(format!("(x={}, y={}, w={}, h={})", x, y, width, height).into())
+}
+
+pub fn create_rectangle_object<'gc>(
+    gc_context: MutationContext<'gc, '_>,
+    rectangle_proto: Option<Object<'gc>>,
+    fn_proto: Option<Object<'gc>>,
+) -> Object<'gc> {
+    FunctionObject::function(
+        gc_context,
+        Executable::Native(constructor),
+        fn_proto,
+        rectangle_proto,
+    )
+}
+
+fn is_empty<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    let width = this
+        .get("width", avm, context)?
+        .resolve(avm, context)?
+        .as_number(avm, context)?;
+    let height = this
+        .get("height", avm, context)?
+        .resolve(avm, context)?
+        .as_number(avm, context)?;
+    Ok((width == 0.0 || height == 0.0 || width.is_nan() || height.is_nan()).into())
+}
+
+fn set_empty<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    this.set("x", 0.into(), avm, context)?;
+    this.set("y", 0.into(), avm, context)?;
+    this.set("width", 0.into(), avm, context)?;
+    this.set("height", 0.into(), avm, context)?;
+    Ok(Value::Undefined.into())
+}
+
+fn clone<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    let proto = context.system_prototypes.rectangle;
+    let args = [
+        this.get("x", avm, context)?.resolve(avm, context)?,
+        this.get("y", avm, context)?.resolve(avm, context)?,
+        this.get("width", avm, context)?.resolve(avm, context)?,
+        this.get("height", avm, context)?.resolve(avm, context)?,
+    ];
+    let cloned = proto.new(avm, context, proto, &args)?;
+    let _ = constructor(avm, context, cloned, &args)?;
+    Ok(cloned.into())
+}
+
+fn contains<'gc>(
```

Comment:
Not really sure how these ones occur:
```
var rect = new Rectangle(1, 1, 1, 1);
trace(rect.contains(false)); // false
trace(rect.contains(true)); // undefined
trace(rect.contains({})); // false
trace(rect.contains(Infinity)); // false
trace(rect.contains(NaN)); // undefined
```
Ruffle returns undefined for them all.

Reply:
Oh dear. Interesting results. The fact that it was a different value for `false` and `true` makes me think it's doing a boolean or maths test. I'll investigate.

Reply:
I looked into this some more and I am stumped. I can't figure out the pattern.

I'm hoping that this would never actually be an issue, unless people were doing `=== false` or `!== true`. I've added a todo comment about it.